Automattic/jetpack
# CRM: Event Manager System - Contact manager

Fixes https://github.com/Automattic/zero-bs-crm/issues/3128

## Proposed changes:

This PR adds the Event Manager system, it's in charge of handling JPCRM events and notifying them via the WP hook system.

This is the first part of the Event Manager system, adding the capabilities for managing contact events. In other PR we will add the rest of the managers (they were removed from this to simplify the PR)

### Other information:


## Jetpack product discussion



## Does this pull request change what data or activity we track or use?



## Testing instructions:
- Check the code
- In `projects/plugins/crm` run `composer dumpautoload`
- Run the unit tests: `composer phpunit -- --testdox --testsuite event-manager``
- Also, you can test it with JPCRM:

Add some actions to your sandbox file: `wp-content/mu-plugins/0-sandbox.php`

```
add_action( 'jpcrm_contact_updated', function( $contact, $old_contact ) {
    error_log( "\nContact Updated!! [{$contact['id']}]", 3, __DIR__ . '/automation.log' );
}, 10, 2 );

add_action( 'jpcrm_contact_created', function( $contact ) {
    error_log( "\nContact Created!! [{$contact['id']}] ", 3, __DIR__ . '/automation.log' );
} );

add_action( 'jpcrm_contact_status_updated', function( $contact, $old_value ) {
    error_log( "\nContact Status updated!! [{$contact['id']}]. Status: $old_value => {$contact['status']}", 3, __DIR__ . '/automation.log' );
}, 10, 2 );

add_action( 'jpcrm_contact_deleted', function( $contact_id ) {
    error_log( "\nContact Deleted!! [$contact_id]", 3, __DIR__ . '/automation.log' );
```
- Now go to JPCRM dashboard, add a contact, update it, and remove it. You should see the log output in a file called `automation.log` in the `mu-plugin` folder.
